src/sandstorm/sandstorm-http-bridge.c++

```diff
@@ -1195,7 +1201,59 @@ private:
     auto req = apiCap.restoreRequest();
     req.setToken(percentDecode(buf));
 
-    return req.send().getCap().castAs<Identity>();
+    // Don't directly call `getCap()` on the `RemotePromise` because we want to make sure
+    // that the returned promise fails if the request fails.
+    return req.send().then([](auto response) {
+      return response.getCap().template castAs<Identity>();
+    });
+  }
+
+  void saveIdentityInternal(kj::StringPtr textId, Identity::Client identity) {
+    auto req = apiCap.saveRequest();
+    req.setCap(identity);
+    req.initLabel().setDefaultText("user identity");
+    tasks.add(req.send().then([this,textId](auto result) -> void {
+      // Sandstorm tokens are primarily text but use percent-encoding to be safe.
+      auto tokenText = percentEncode(result.getToken());
+
+      // Clean up any existing symlink.
+      dropIdentity(textId);
+
+      // Store as a symlink. ext4 can store up to 60 bytes directly in the inode, avoiding
+      // allocating a block.
+      KJ_SYSCALL(symlinkat(tokenText.cStr(), identitiesDir, textId.cStr()));
+
+      // Make sure it's really saved.
+      KJ_SYSCALL(fsync(identitiesDir));
+    }));
+  }
+
+  void dropIdentity(kj::StringPtr textId) {
+    auto symlink = kj::heapString(textId);
+
+    if (faccessat(identitiesDir, symlink.cStr(), F_OK, AT_SYMLINK_NOFOLLOW) == 0) {
+      char buf[512];
+      ssize_t n;
+      KJ_SYSCALL(n = readlinkat(identitiesDir, symlink.cStr(), buf, sizeof(buf)));
+      KJ_ASSERT(n < sizeof(buf), "token too long?");
+      buf[n] = '\0';
+
+      // We name the trash file after the token, not the identity ID. This way, it's okay
+      // if we overwrite an existing entry of the trash directory.
+      auto trashSymlink = kj::heapString(buf);
+      KJ_SYSCALL(renameat(identitiesDir, symlink.cStr(), trashDir, trashSymlink.cStr()));
+
+      auto req = apiCap.dropRequest();
+      req.setToken(percentDecode(buf));
+      tasks.add(req.send().then([KJ_MVCAP(trashSymlink), this](auto response) -> void {
+        KJ_SYSCALL(unlinkat(trashDir, trashSymlink.cStr(), 0));
+        KJ_SYSCALL(fsync(trashDir));
```

Comment:
I think this fsync() isn't accomplishing anything except to slow things down, since nothing is actually gated on the completion of this operation. Normally you need to fsync() before returning a success code to the caller, so that the caller knows that success means durability, but there's no caller here.

With or without the fsync(), it's possible to be left in a state where there are tokens in the trash that were in fact already dropped. That's fine, since the worst case is that drop() will be harmlessly called again later.